# 5. Funciones de Ventana para el Análisis de Datos

In [1]:
# @title Postgress preloading (install server)
#The output of the installation is not displayed when %%capture is used at the start of the cell
%%capture
# Install postgresql server
!sudo apt-get -y -qq update
!sudo apt-get -y -qq install postgresql
!sudo service postgresql start
# Setup a password `postgres` for username `postgres`
!sudo -u postgres psql -U postgres -c "ALTER USER postgres PASSWORD 'postgres';"
# Setup a database with name `sampledb` to be used
# !sudo -u postgres psql -U postgres -c 'DROP DATABASE IF EXISTS sampledb;'
# !sudo -u postgres psql -U postgres -c 'CREATE DATABASE sampledb;'

!curl https://raw.githubusercontent.com/limspiga/data-modeling/main/db/data.dump  -O
!sudo -u postgres psql -U postgres -c 'CREATE DATABASE sqlda;'

# !sudo -u postgres psql -U postgres -d 'sqlda' -f 'data.dump'

# https://thivyapriyaa.medium.com/setting-up-postgresql-on-google-colab-4d02166939fc


In [2]:
# @title Postgress preloading
# import
!sudo -u postgres psql -d sqlda < data.dump
%env DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sqlda
#To load the sql extention to start using %%sql
%load_ext sql

SET
SET
SET
SET
SET
 set_config 
------------
 
(1 row)

SET
SET
SET
SET
CREATE EXTENSION
COMMENT
CREATE EXTENSION
COMMENT
CREATE TEXT SEARCH DICTIONARY
ALTER TEXT SEARCH DICTIONARY
SET
SET
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE MATERIALIZED VIEW
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE MATERIALIZED VIEW
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
COPY 44533
COPY 0
COPY 50000
COPY 32
COPY 50000
COPY 20
COPY 418158
COPY 12
COPY 15412
COPY 37711
COPY 300
COPY 20
ALTER TABLE
ALTER TABLE
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
REFRESH MATERIALIZED VIEW
REFRESH MATERIALIZED VIEW
env: DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sqlda


En el Capítulo 3, aprendiste funciones de SQL para la preparación de datos, como CASE WHEN, COALESCE y NULLIF, que actúan sobre datos individuales de una fila. Por otro lado, el Capítulo 4 te introdujo a funciones de agregación, como SUM, AVG y COUNT, que analizan conjuntos de filas. Un ejemplo práctico es el análisis de medidas de recién nacidos; podrías usar CASE WHEN para verificar la salud individual basándote en el peso y la altura, y funciones de agregación para determinar promedios y desviaciones estándar a nivel nacional.

A veces, la posición relativa de un dato en un conjunto es esencial, como en el caso de los rangos. Un bebé podría tener un rango diferente en medidas de salud a nivel nacional y a nivel de ciudad. Además, en un conjunto grande, puedes querer analizar subgrupos o particiones, como estados. Para ello, se calcula un rango dentro de cada partición, considerando un conjunto seleccionado de filas o "ventana". Esta forma de cálculo se logra con lo que se denomina función de ventana.


# Funciones de Ventana
Siguiendo con la discusión sobre las Funciones de Ventana, deseas encontrar los primeros clientes de ZoomZoom. Técnicamente, esto significa que quieres clasificar a cada cliente según la fecha en que se convirtieron en clientes, siendo el primero en el rango 1, el siguiente en el rango 2 y así sucesivamente. Puedes obtener todos los clientes usando la siguiente consulta:

In [4]:
%%sql
-- Figure 5.1: Customers ordered by date_added
SELECT
  customer_id, first_name, last_name, date_added
FROM
  customers
ORDER BY
  date_added LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(18685, 'Ingram', 'Crossman', datetime.datetime(2012, 11, 9, 0, 0)),
 (30555, 'Stanwood', 'Stookes', datetime.datetime(2012, 11, 9, 0, 0)),
 (17099, 'Pearla', 'Halksworth', datetime.datetime(2012, 11, 9, 0, 0)),
 (6173, 'Danila', 'Gristwood', datetime.datetime(2012, 11, 9, 0, 0)),
 (12484, 'Lillis', 'Brayley', datetime.datetime(2012, 11, 9, 0, 0)),
 (30046, 'Nanete', 'Hassur', datetime.datetime(2012, 11, 9, 0, 0)),
 (13390, 'Danika', 'Lough', datetime.datetime(2012, 11, 9, 0, 0)),
 (7486, 'Ciro', 'Ferencowicz', datetime.datetime(2012, 11, 9, 0, 0)),
 (2625, 'Binky', 'Dawtrey', datetime.datetime(2012, 11, 9, 0, 0)),
 (35683, 'Betteanne', 'Rulf', datetime.datetime(2012, 11, 9, 0, 0))]

Puedes ordenar a los clientes desde el más antiguo hasta el más reciente, copiar los resultados en una hoja de Excel y asignar un número de fila a cada uno para tener el rango de cada cliente. Sin embargo, este método no es automático y puede llevar a errores. SQL ofrece varias formas para lograr esto de manera más eficiente. Más adelante en este capítulo, aprenderás cómo asignar números a registros ordenados usando la función de ventana RANK. Aquí, puedes comenzar usando una función de agregación para obtener las fechas y ordenarlas de esa manera.

In [6]:
%%sql
-- Figure 5.2: Aggregate date-time ordering
SELECT
  date_added, COUNT(*)
FROM
  customers
GROUP BY
  date_added
ORDER BY
  date_added LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(datetime.datetime(2012, 11, 9, 0, 0), 11),
 (datetime.datetime(2012, 11, 10, 0, 0), 13),
 (datetime.datetime(2012, 11, 11, 0, 0), 12),
 (datetime.datetime(2012, 11, 12, 0, 0), 19),
 (datetime.datetime(2012, 11, 13, 0, 0), 23),
 (datetime.datetime(2012, 11, 14, 0, 0), 16),
 (datetime.datetime(2012, 11, 15, 0, 0), 20),
 (datetime.datetime(2012, 11, 16, 0, 0), 14),
 (datetime.datetime(2012, 11, 17, 0, 0), 11),
 (datetime.datetime(2012, 11, 18, 0, 0), 21)]

El resultado ordena las fechas de manera secuencial. A partir de ello, es posible determinar cuántos clientes se unieron a ZoomZoom antes que un cliente específico, sumando las inscripciones de los días anteriores a su fecha de adhesión. Sin embargo, este enfoque es manual y no ofrece una visión directa del rango. Es aquí donde entran en juego las funciones de ventana. Estas herramientas procesan varias filas de datos conservando toda su información. Son esenciales para tareas como determinar rangos. En la siguiente sección, ilustraremos con un ejemplo de consulta utilizando una función de ventana.

# Conceptos Básicos de las Funciones de Ventana
A continuación, se presenta la sintaxis básica de una función de ventana:

```sql
SELECT {columns},
{window_func} OVER (PARTITION BY {partition_key} ORDER BY {order_key})
FROM table1;
```

Aquí, {columnas} son las columnas que deseas recuperar de las tablas para la consulta, {func_ventana} es la función de ventana que deseas usar, table1 es la tabla o tablas unidas de las que deseas extraer datos, y la palabra clave OVER indica donde comienza la definición de la ventana. Esta definición, en su sintaxis básica, incluye dos partes: {clave_partición} y {clave_orden}. La primera se refiere a la columna o columnas que deseas dividir (se hablará más de esto más adelante) y la segunda es la columna o columnas por las que deseas ordenar.

Para ilustrar esto, consideremos un ejemplo. Podrías pensar que no conoces ninguna función de ventana, pero la realidad es que todas las funciones de agregación pueden utilizarse como funciones de ventana. Ahora, utiliza COUNT(*) en la siguiente consulta:

In [9]:
%%sql
-- Figure 5.3: Customers listed using the COUNT(*) window query
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER () as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 50000),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 50000),
 (3, None, 'Braden', 'Jordan', 'M', 50000),
 (4, None, 'Jessika', 'Nussen', 'F', 50000),
 (5, None, 'Lonnie', 'Rembaud', 'F', 50000),
 (6, None, 'Cortie', 'Locksley', 'M', 50000),
 (7, None, 'Wood', 'Kennham', 'M', 50000),
 (8, None, 'Rutger', 'Humblestone', 'M', 50000),
 (9, None, 'Melantha', 'Tibb', 'F', 50000),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 50000)]

Como se muestra en la Figura 5.3, la consulta devuelve título, primer nombre y apellido, al igual que una consulta SELECT típica. Sin embargo, ahora hay una nueva columna llamada total_clientes. Esta columna contiene el recuento de usuarios que serían creados por la siguiente consulta:

In [11]:
%%sql
SELECT
  COUNT(*)
FROM
  customers;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(50000,)]

La consulta anterior devuelve 50,000. La consulta retornó todas las filas y el COUNT(*) en la consulta devuelve el recuento como lo haría cualquier función de agregación normal.
Ahora, en cuanto a los otros parámetros de la consulta, ¿qué sucede si añades la cláusula OVER para convertir este COUNT en una función de ventana, manteniendo la función como COUNT pero definiendo la ventana usando PARTITION BY, como en la consulta siguiente?

In [13]:
%%sql
-- Figure 5.4: Customers listed using COUNT(*) partitioned by the gender window query
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER (PARTITION BY gender) as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 25044),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 24956),
 (3, None, 'Braden', 'Jordan', 'M', 24956),
 (4, None, 'Jessika', 'Nussen', 'F', 25044),
 (5, None, 'Lonnie', 'Rembaud', 'F', 25044),
 (6, None, 'Cortie', 'Locksley', 'M', 24956),
 (7, None, 'Wood', 'Kennham', 'M', 24956),
 (8, None, 'Rutger', 'Humblestone', 'M', 24956),
 (9, None, 'Melantha', 'Tibb', 'F', 25044),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 25044)]

Aquí, puedes ver que total_clientes ahora ha cambiado los conteos a uno de dos valores, 24956 o 25044. Al usar la cláusula PARTITION BY sobre la columna de género, SQL divide el conjunto de datos en múltiples particiones basadas en los valores únicos de esta columna. Dentro de cada partición, SQL calcula el conteo total. Por ejemplo, hay 24956 hombres, por lo que la función de ventana COUNT para la partición de hombres devuelve 24596, lo que puedes confirmar con la siguiente consulta:

In [14]:
%%sql
SELECT
  gender,
  COUNT(*)
FROM
  customers
GROUP BY
1;

 * postgresql://postgres:***@localhost:5432/sqlda
2 rows affected.


[('M', 24956), ('F', 25044)]

Ahora ves cómo se define y se utiliza la partición con la cláusula PARTITION BY. Para las mujeres, el conteo es igual al número de mujeres, y para los hombres, el conteo es igual al número de hombres. ¿Qué sucede ahora si utilizas ORDER BY en lugar de ello en la cláusula OVER como sigue?

In [17]:
%%sql
-- Figure 5.5: Customers listed using COUNT(*) ordered by the customer_id window query
SELECT
  customer_id, title,
  first_name, last_name, gender,
  COUNT(*) OVER (ORDER BY customer_id) as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 2),
 (3, None, 'Braden', 'Jordan', 'M', 3),
 (4, None, 'Jessika', 'Nussen', 'F', 4),
 (5, None, 'Lonnie', 'Rembaud', 'F', 5),
 (6, None, 'Cortie', 'Locksley', 'M', 6),
 (7, None, 'Wood', 'Kennham', 'M', 7),
 (8, None, 'Rutger', 'Humblestone', 'M', 8),
 (9, None, 'Melantha', 'Tibb', 'F', 9),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 10)]

Notarás algo similar a un conteo acumulado para el total de clientes. Aquí es donde proviene la definición de 'ventana' en función de ventana. Cuando usas esta función de ventana, dado que no especificaste un PARTITION BY, se utiliza el conjunto de datos completo para el cálculo. Dentro de este conjunto de datos, cuando no se especifica ORDER BY, se asume que solo hay una ventana, que contiene todo el conjunto de datos. Sin embargo, cuando se especifica ORDER BY, las filas en el conjunto de datos se ordenan de acuerdo con este criterio. Para cada valor único en el orden, SQL forma un grupo de valores, que contiene todas las filas con dicho valor. La consulta luego crea una ventana para cada grupo de valores. La ventana contendrá todas las filas en este grupo de valores y todas las filas que están ordenadas antes de este grupo de valores. A continuación se muestra un ejemplo:

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/8daed639-c93e-4a0c-a7d4-27c16ee9c892.png?raw=true' width="600" />
<figcaption>
Figure 5.6: Windows for customers using COUNT(*) ordered by the customer_id window query</figcaption></center>
</figure>

Aquí, el conjunto de datos se ordena usando customer_id, que resulta ser la clave primaria. Como tal, cada fila tiene un valor único y forma un grupo de valores. El primer grupo de valores, sin ninguna fila antes que él, forma su propia ventana, que contiene solo la primera fila. La ventana del segundo grupo de valores contendrá tanto a sí misma como a la fila anterior, lo que significa la primera y segunda fila. Luego, la ventana del tercer grupo de valores contendrá a sí misma y las dos filas anteriores, y así sucesivamente. Cada grupo de valores tiene su ventana. Una vez que se establecen las ventanas, para cada grupo de valores, se calcula la función de ventana basada en la ventana. En este ejemplo, esto significa que COUNT se aplica a cada ventana. Por lo tanto, el grupo de valores 1 (la primera fila) obtiene 1 como resultado ya que su Ventana 1 contiene una fila, el grupo de valores 2 (la segunda fila) obtiene 2 ya que su Ventana 2 contiene dos filas, y así sucesivamente. Los resultados se aplican a cada fila en este grupo de valores si el grupo contiene varias filas. Ten en cuenta que la ventana se utiliza solo para el cálculo. Los resultados se asignan a filas en el grupo de valores, no se asignan a las filas en la ventana.

¿Qué sucede cuando combinas PARTITION BY y ORDER BY? Ahora, observa la siguiente consulta:

In [20]:
%%sql
-- Figura 5.7: Clientes listados usando COUNT(*) particionados por género y ordenados por la consulta de ventana customer_id.
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER (
PARTITION BY gender ORDER BY customer_id
  ) as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 1),
 (3, None, 'Braden', 'Jordan', 'M', 2),
 (4, None, 'Jessika', 'Nussen', 'F', 2),
 (5, None, 'Lonnie', 'Rembaud', 'F', 3),
 (6, None, 'Cortie', 'Locksley', 'M', 3),
 (7, None, 'Wood', 'Kennham', 'M', 4),
 (8, None, 'Rutger', 'Humblestone', 'M', 5),
 (9, None, 'Melantha', 'Tibb', 'F', 4),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 5)]

Al igual que la consulta anterior, parece ser una especie de rango. Sin embargo, parece diferir según el género. En este SQL en particular, la consulta divide la tabla en dos subconjuntos basados en la columna PARTITION BY. Eso es porque la cláusula PARTITION BY, al igual que GROUP BY, primero divide el conjunto de datos en grupos (que aquí se llama partición) basados en el valor en la columna de género. Luego, cada partición se utiliza como base para hacer un conteo, con cada partición teniendo su propio conjunto de grupos de valores. Estos grupos de valores se ordenan dentro de la partición, se crean ventanas basadas en los grupos de valores y sus órdenes, y la función de ventana se aplica a los valores. Los resultados se asignan finalmente a cada fila en los grupos de valores.


Este proceso se ilustra en la Figura 5.8. Este proceso produce el conteo que puedes ver. Las tres palabras clave, OVER(), PARTITION BY y ORDER BY, son la base del poder de las funciones de ventana.

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/6a82dd5c-88c2-4fec-bd7e-228233ed5585.png?raw=true' width="600" />
<figcaption>
Figura 5.8: Ventanas para clientes listados usando COUNT(*) particionados por género y ordenados por la consulta de ventana customer_id.</figcaption></center>
</figure>

Ahora que comprendes las funciones de ventana, intenta aplicarlas en el próximo ejercicio.

## Ejercicio 5.01: Analizando tasas de llenado de datos de clientes a lo largo del tiempo.

En este ejercicio, aplicarás funciones de ventana a un conjunto de datos y analizarás la información. Durante los últimos seis meses, ZoomZoom ha estado experimentando con diversas promociones para lograr que sus clientes estén más involucrados en la actividad de venta. Una forma de medir el nivel de compromiso es evaluar la disposición de las personas para completar todos los campos en el formulario de cliente, especialmente su dirección. Para lograr este objetivo, la empresa desea un total acumulado de cuántos usuarios han rellenado sus direcciones postales con el tiempo. Escribe una consulta para producir estos resultados.